go/storage/mkvs/db/pathbadger/pathbadger.go

```diff
@@ -0,0 +1,943 @@
+				if seqNo > 0 {
+					continue // All nodes with higher seqNo will be cleared anyway.
+				}
+
+				maybeLoneNodes[rht][string(un.Key)] = struct{}{}
+			}
+
+			// Remove write logs for the non-finalized root.
+			if !d.discardWriteLogs {
+				if err = func() error {
+					rootWriteLogsPrefix := writeLogKeyFmt.Encode(version, &rootHash)
+					wit := tx.NewIterator(badger.IteratorOptions{Prefix: rootWriteLogsPrefix})
+					defer wit.Close()
+
+					for wit.Rewind(); wit.Valid(); wit.Next() {
+						if err = batchMeta.Delete(wit.Item().KeyCopy(nil)); err != nil {
+							return err
+						}
+					}
+					return nil
+				}(); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	rootIt.Close()
+
+	// Saniy check that all finalized roots were visited.
+	if nonEmptyFinalizedRoots != nonEmptyVisitedRoots {
+		return fmt.Errorf("mkvs/pathbadger: not all finalized roots were visited (db corruption?)")
+	}
+
+	// Copy over any updated nodes for finalized roots with non-zero sequence numbers.
+	for rht, nodes := range notLoneNodes {
+		seqNo := finalizedSeqNos[rht]
+		if seqNo == 0 {
+			continue // Already in the right place.
+		}
+
+		for k := range nodes {
+			// Fetch pending node value. We will remove it later.
+			item, err := tx.Get(pendingNodeKeyFmt.Encode(version, rht, seqNo, []byte(k)))
+			if err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to copy node: %w", err)
+			}
+			var value []byte
+			err = item.Value(func(data []byte) error {
+				value = append([]byte{}, data...) // Must copy to queue in batch.
+				return nil
+			})
+			if err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to copy node: %w", err)
+			}
+
+			// Copy over to new location.
+			if err := batch.Set(finalizedNodeKeyFmt.Encode(rht, []byte(k)), value); err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to copy node: %w", err)
+			}
+		}
+	}
+	// Remove any lone nodes.
+	for rht, nodes := range maybeLoneNodes {
+		for k := range nodes {
+			if _, isNotLone := notLoneNodes[rht][k]; isNotLone {
+				continue
+			}
+
+			if err := batch.Delete(finalizedNodeKeyFmt.Encode(rht, []byte(k))); err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to delete lone node: %w", err)
+			}
+		}
+	}
+
+	// Remove all temporary nodes for non-zero sequence numbers.
+	pendingPrefix := pendingNodeKeyFmt.Encode(version)
+	pendingIt := tx.NewIterator(badger.IteratorOptions{Prefix: pendingPrefix})
+	defer pendingIt.Close()
+
+	for pendingIt.Rewind(); pendingIt.Valid(); pendingIt.Next() {
+		if err := batchMeta.Delete(pendingIt.Item().KeyCopy(nil)); err != nil {
+			return fmt.Errorf("mkvs/pathbadger: failed to delete pending node: %w", err)
+		}
+	}
+
+	pendingIt.Close()
+
+	// Remove any queued keys. Do this last as this action is irreversible in case of crash.
+	for _, key := range removeMetaKeys {
+		if err := batchMeta.Delete(key); err != nil {
+			return fmt.Errorf("mkvs/pathbadger: failed to delete key: %w", err)
+		}
+	}
+
+	// Commit batches.
```

Comment:
Not sure what happens if `batch` commit or `batchMeta` commit fails. It would be nice to have a comment that explains this.